proposals/generic-ref-struct.md

```diff
@@ -0,0 +1,424 @@
+
+The compiler will introduce the type `System.Runtime.CompilerServices.SupportsOnlyNonByRefLikeAttribute` that can be applied to methods and operators. When applied to a member it has the effect of removing `where T : ref struct` from the constraints on all type parameters on the *immediate* containing type. All other constraints remain in place.
+
+This allows us to rationalize our `Span<T>` definition as well as allowing third party customers rationalize their existing APIs. 
+
+```c#
+// Existing Span<T> definition
+public readonly ref struct Span<T>
+{
+    [SupportsOnlyNonByRefLikeAttributes]
+    public Span(T[]? array) { ... }
+
+    [SupportsOnlyNonByRefLikeAttributes]
+    public Span(T[]? array, int index, int length) { ... }
+
+    [SupportsOnlyNonByRefLikeAttributes]
+    public static implicit operator Span<T>(T[]? array) { ... }
+}
+```
+
+This does not impact the binding rules for invocations of such members. These symbols will continue to be fully bindable. In the case they are bound when any of the type parameters on the containing type are potentially a `ref struct` then use of a member annotated with `[SupportsOnlyNonByRefLike]` will be considered an error.
+
+```c#
+void M<T, U>()
+    where T : ref struct
+{
+    // Error: cannot access member as `T` is potentially a `ref struct`
+    new Span<T>(null);
+
+    // Okay
+    new Span<U>(null);
+}
+```
+
+The runtime will recognize this attribute and ensure that any use of these members at runtime with a type that is a `ref struct` will result in a runtime exception.
+
+Note: the [alternative](#special-case) here is simply special casing `Span<T>` as array is special cased today.
+
+### ref struct interfaces
+<a name="interfaces"></a>
+
+The language will allow `ref struct` to implement interfaces. The rules for implementation will be the same as for normal `struct`. This includes allowing the implementations to be implicit or explicit.
+
+**TODO: Need better example than animals but feeling lazy**
+
+```c#
+interface IAnimal
+{
+    void Speak(); 
+}
+
+ref struct Lab : IAnimal
+{
+    public void Speak() => Console.WriteLine("FEED ME");
+}
+```
+
+The one exception is when `ref struct` intersects with default implementation methods. When interfaces containing default implemented methods are used on `struct` types there is an implicit boxing operation for the invocation. That means effectively a DIM call when the implementing type was a `ref struct` would violate our rules on them appearing in the heap. As such this will be disallowed by both the language and runtime. 
+
+- A `ref struct` that implements an `interface` must provide implementations for all methods, even those with a default implementation
```

Comment:
An important motivation for DIMs is they allow developers to evolve interfaces without breaking implementors.

Forcing ref structs to implement DIMs means that adding a DIM will break any ref structs that implement the interface.

Are we ok saying that that's rare enough it shouldn't be a concern to interface authors?

Alternatively should interfaces have to opt into being implemented by ref structs, or have the ability to opt out?

Reply:
> Are we ok saying that that's rare enough it shouldn't be a concern to interface authors?

I tend to think of it in the opposite direction. Today there is no way to implement an interface on a `ref struct` hence we are starting from scratch here (no compat concerns). To me the cost of implementing an interface on a `ref struct` becomes that your implementation will not benefit from DIM. Basically you must manually implement all interface members. 

 > Alternatively should interfaces have to opt into being implemented by ref structs, or have the ability to opt out?

I think forcing interfaces to opt-into `ref struct` would unnecessarily slow down the ecosystem. Every time a developer wanted to adapt an `ref struct` to an `interface`, even if it was for a limited use case, they'd have to get the developer to opt the interface in, go through an update cycle, etc ... Very slow. 

As for opting an interface out of `ref struct` implementations I'm not sure of the value. The only advantage I could see that providing is it gives developers the feeling that they could add DIM without breaking people. Problem is that is just not a reality. Due to diamond implementation problems there is always a risk to adding a DIM to an existing shipped interface. The risk depends on a number of factors but in many cases it's non-zero (it's not just a .NET thing here, Java has the same issues). 

This is a case where I'd say "probably better as an analyzer" if you really wanted it.